### PR TITLE
Use new registers which give the state of the configuration process during setup

### PR DESCRIPTION
**Pull request by swh76**
## Issue
This PR resolves issue https://github.com/slaclab/pysmurf/issues/462.

## Description

This PR mostly expands the `smurf_command.set_defaults_pv` routine to check if loading the hardware defaults succeeded using the new `ConfiguringInProgress` and `SystemConfigured` registers described in https://github.com/slaclab/pysmurf/issues/462.  These new registers are not provided in the current stable pysmurf release (v4.0.0) so to use them right now, you have to run with a dev_sw rogue docker with the below version of pysmurf.  Here's an attempt at a comprehensive list of changes made:

* Added new `get_configuring_in_progress` and `get_system_configured` routines to `smurf_command` to poll the new `ConfiguringInProgress` and `SystemConfigured` registers.
* Adjusts the RF lab configuration files to use the newer dev_sw version I tested this code on.
* Elevated the wait after the `setDefaults` calls in `set_defaults_pv` to a new `set_defaults_pv` routine keyword `wait_after_sec`.
* Added configuration success check suggested in issue https://github.com/slaclab/pysmurf/issues/462.  The check is only done if the pysmurf core code version is >=4.1.0.
* shawnhammer now checks if `setup` fails.
* `setup` and `set_defaults_pv` now return a boolean success flag which is False if they fail, True if they succeed.  `set_defaults_pv` returns None for pysmurf core code versions <=4.1.0.
* Two new keywords `max_timeout_sec` and `caget_timeout_sec` added to the `set_defaults_pv` routine to allow user to control how long the routine waits for the system to configure/respond to register polls.  
* Adds Shawn's rf_deterministic_latency.py test script.
* Expands docstrings for the `setup` and `set_defaults_pv` routines.  The `setup` routine still needs more work.  I've created a new pysmurf issue for that here - https://github.com/slaclab/pysmurf/issues/475.

Tested with these software versions:

smurf_cfg : v1.3.0
carrier fw : ae09de6
Rogue zip : v0.1.10
pysmurf(master) : fcf349ac
Rogue/pyrogue version : v4.10.6

## Does this PR break any interface?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by swh76** [comment]
@eyyoung24 @jesusvasquez333 @ruck314 I think this works but I found I have to put in a 30 sec sleep between when ConfiguringInProgress goes to False after setDefaults and when SystemConfigured goes to True (if successful).  It's probably less than 30 sec, but 10 sec didn't work so I bumped it up to 30 sec to be safe.

https://github.com/slaclab/pysmurf/blob/24451421a3c6ec4f1f5057bd8ca0a32a63b92b59/python/pysmurf/client/command/smurf_command.py#L596

It also looks like SystemConfigured might not get set to False if you run setDefaults a second time, which may or may not be the behavior we want.

Is there anyway to have ConfiguringInProgress wait to go False until SystemConfigured is True?

**Comment by jesusvasquez333** [comment]
@swh76 I tested your code in Lab2. I removed:
- The version checking:
https://github.com/slaclab/pysmurf/blob/24451421a3c6ec4f1f5057bd8ca0a32a63b92b59/python/pysmurf/client/command/smurf_command.py#L544
- The 30s delay
https://github.com/slaclab/pysmurf/blob/24451421a3c6ec4f1f5057bd8ca0a32a63b92b59/python/pysmurf/client/command/smurf_command.py#L596

With these changes it works as expected. I verified using the GUI as well. This is what I saw:
1. Started a new server + client. Both `ConfiguredInProgress` and `SystemConfigured` are set to `False` initially.
Screen Shot 2020-07-23 at 10 26 14 AM

2. Call the `set_defaults_pvs()` command. 
```
In [1]: S = pysmurf.client.SmurfControl(epics_root=epics_prefix,cfg_file=config_file,setup=False,make_logfile=False,shelf_manager="shm-smrf-sp01")

In [2]: S.set_defaults_pv(write_log=True)
```
`ConfiguredInProgress` changes to `True`. `SystemConfigured` remains `False`. 
Screen Shot 2020-07-23 at 10 26 24 AM

3. When the process ends, `ConfiguredInProgress` changes to `False`, and `SystemConfigured` changes to `True` at the same time.
Screen Shot 2020-07-23 at 10 27 12 AM
 
The final command log is:
```
In [2]: S.set_defaults_pv(write_log=True)
[ 2020-07-23 17:26:20 ]  caget smurf_server_s2:AMCc:SmurfApplication:SmurfVersion
[ 2020-07-23 17:26:20 ]  4.0.0+568.g24451421.dirty
[ 2020-07-23 17:26:20 ]  caput smurf_server_s2:AMCc:setDefaults 1
CA.Client.Exception...............................................
    Warning: "Virtual circuit unresponsive"
    Context: "localhost:43783"
    Source File: ../tcpiiu.cpp line 920
    Current Time: Thu Jul 23 2020 17:26:55.968652099
..................................................................
[ 2020-07-23 17:26:55 ]  caget smurf_server_s2:AMCc:SmurfApplication:ConfiguringInProgress
cannot connect to smurf_server_s2:AMCc:SmurfApplication:ConfiguringInProgress
[ 2020-07-23 17:27:05 ]  None
[ 2020-07-23 17:27:05 ]  caget smurf_server_s2:AMCc:SmurfApplication:SystemConfigured
[ 2020-07-23 17:27:08 ]  1
[ 2020-07-23 17:27:08 ]  System configuration finished after 47 seconds. The final state was True.
Out[2]: True
```
It returns `True` which is correct.

4. Call the `set_defaults_pvs()` one more time. 
```
In [3]: S.set_defaults_pv(write_log=True)
```
`ConfiguredInProgress` changes to `True`. `SystemConfigured` changes `False`:
Screen Shot 2020-07-23 at 10 28 43 AM

5.  When the process ends, `ConfiguredInProgress` changes to `False`, and `SystemConfigured` changes to `True` at the same time (as before).
Screen Shot 2020-07-23 at 10 29 31 AM

The final system command logs is the same as before:
```
In [3]: S.set_defaults_pv(write_log=True)
[ 2020-07-23 17:28:39 ]  caget smurf_server_s2:AMCc:SmurfApplication:SmurfVersion
[ 2020-07-23 17:28:39 ]  4.0.0+568.g24451421.dirty
[ 2020-07-23 17:28:39 ]  caput smurf_server_s2:AMCc:setDefaults 1
CA.Client.Exception...............................................
    Warning: "Virtual circuit unresponsive"
    Context: "localhost:43783"
    Source File: ../tcpiiu.cpp line 920
    Current Time: Thu Jul 23 2020 17:29:14.758726324
..................................................................
[ 2020-07-23 17:29:14 ]  caget smurf_server_s2:AMCc:SmurfApplication:ConfiguringInProgress
cannot connect to smurf_server_s2:AMCc:SmurfApplication:ConfiguringInProgress
[ 2020-07-23 17:29:24 ]  None
[ 2020-07-23 17:29:24 ]  caget smurf_server_s2:AMCc:SmurfApplication:SystemConfigured
[ 2020-07-23 17:29:26 ]  1
[ 2020-07-23 17:29:26 ]  System configuration finished after 46 seconds. The final state was True.
Out[3]: True
```
It returns `True` again.

6. **I modified the defaults file and added a wrong value in order to make the configuration process to fail.**  Call the `set_defaults_pvs()` one more time. 
```
In [4]: S.set_defaults_pv(write_log=True)
```
`ConfiguredInProgress` changes to `True`. `SystemConfigured` changes `False`:
Screen Shot 2020-07-23 at 10 30 10 AM

7. When the process ends, `ConfiguredInProgress` changes to `False`, and `SystemConfigured` remains `False`:
Screen Shot 2020-07-23 at 10 30 36 AM

The final system command logs is:
```
In [4]: S.set_defaults_pv(write_log=True)
[ 2020-07-23 17:30:08 ]  caget smurf_server_s2:AMCc:SmurfApplication:SmurfVersion
[ 2020-07-23 17:30:08 ]  4.0.0+568.g24451421.dirty
[ 2020-07-23 17:30:08 ]  caput smurf_server_s2:AMCc:setDefaults 1
[ 2020-07-23 17:30:10 ]  caget smurf_server_s2:AMCc:SmurfApplication:ConfiguringInProgress
[ 2020-07-23 17:30:10 ]  0
[ 2020-07-23 17:30:10 ]  caget smurf_server_s2:AMCc:SmurfApplication:SystemConfigured
[ 2020-07-23 17:30:10 ]  0
[ 2020-07-23 17:30:10 ]  System configuration finished after 2 seconds. The final state was False.
Out[4]: False
```

It returns `False` which is correct.

File: python/pysmurf/client/command/smurf_command.py [hunk 1]
```diff
@@ -408,12 +487,132 @@ def get_number_processed_channels(self, band=None, **kwargs):
...
+        --------
+        :func:`get_configuring_in_progress` : Whether or not
+                configuration process in progress.
+        :func:`get_system_configured` : Returns final state of
+                configuration process.
+
+        """
+        # strip any commit info off the end of the pysmurf version
+        # string
+        pysmurf_version = self.get_pysmurf_version(**kwargs).split('+')[0]
```
Comment (jesusvasquez333) [inline python/pysmurf/client/command/smurf_command.py:538]: @swh76 why do you read the pysmurf version register each time this command is call? Shouldn't this be done the the `pysmurf.client.SmurfControl` is instantiated, and assumes it remains the same? 

Unless you are planning to support restarting the rogue server and keep the same pysmurf client session running?

**Comment by swh76** [comment]
Talking with @jesusvasquez333 I realized my first testing where I needed a 30sec time out between when ConfiguringInProgress goes to False after setDefaults and when SystemConfigured goes to True (if successful) was missing your caget_timeout for the ConfiguringInProgress polling which I later added.  Not seeing the issue now, will remove that 30 sec wait.

File: python/pysmurf/client/command/smurf_command.py (hunk 1 above)
Comment (swh76) [reply python/pysmurf/client/command/smurf_command.py:538]: That is why but I agree caching is smarter.  I'll add something for that ; wait to push.

Comment (swh76) [reply python/pysmurf/client/command/smurf_command.py:538]: @jesusvasquez333 ok sounds good, I'll leave as is, at least for now.  I started a new issue to think about this if we want to revisit in the future - https://github.com/slaclab/pysmurf/issues/476.

**Pull request merged**